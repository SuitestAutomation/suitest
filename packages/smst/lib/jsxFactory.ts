// Reference is needed so that whenever this file is imported by consumer, intrinsic types are always included
/// <reference path="../types/intrinsicElements.d.ts" />
import ub from 'unist-builder';
import {
	CodeBlockNode,
	ConditionNode,
	InlineTextNode,
	Node,
	PropertyNode,
	SingleNode,
	TestLineNode,
	TextNode,
	InlinePropertyNode,
	CodePropertyNode,
} from '../types/unistTestLine';

const plainTypes = ['text', 'code', 'subject', 'input'];
const isTextNode = (input?: JSX.SmstFlatNode): input is TextNode =>
	typeof input === 'object' && input !== null && plainTypes.includes(input.type);

type DeepArrayOrOne<T> = T | Array<T | DeepArrayOrOne<T>>;
export const flatten = <T extends any>(input: DeepArrayOrOne<T>): T[] => {
	if (Array.isArray(input)) {
		return input.reduce((acc: T[], item: DeepArrayOrOne<T>) => acc.concat(flatten(item)), []);
	}

	return [input];
};

/**
 * Normalize textual children - make them flat and merge sequential text items of same type
 */
function normalizePlainChildren
	<R extends SingleNode, K extends keyof JSX.IntrinsicElements = R['type']>
	(children: JSX.SmstFlatNode[], parentType: K): R[];
function normalizePlainChildren(
	children: JSX.SmstFlatNode[],
	parentType: keyof JSX.IntrinsicElements
): SingleNode[] {
	return children.reduce((output: SingleNode[], child) => {
		if (child === undefined || child === null) {
			return output;
		}

		if (typeof child === 'string' || typeof child === 'boolean' || typeof child === 'number') {
			// Wrap plain strings into text nodes
			child = ub(plainTypes.includes(parentType) ? parentType : 'text', String(child)) as InlineTextNode;
		}

		// Nested child is a textual node
		const lastChild = output[output.length - 1];

		if (isTextNode(child) && lastChild?.type === child.type) {
			// Textual node with same time as the previous one - merge them
			output.splice(output.length - 1, 1, {
				type: child.type,
				value: lastChild.value + child.value,
			});
		} else {
			output.push(child);
		}

		return output;
	}, []);
}

const processTextNode = (label: JSX.SmstNode): TextNode[] =>
	normalizePlainChildren<TextNode>(flatten([label]), 'text');

const processPropertyNode = (props: JSX.ElementsProps['prop']): InlinePropertyNode | CodePropertyNode => {
	const name = processTextNode(props.name);

	if (isCodeBlockNode(props.expectedValue)) {
		return ub('prop', {
			name,
			contentType: 'block',
			expectedValue: props.expectedValue,
			comparator: props.comparator,
			status: props.status,
		});
	}

	return ub('prop', {
		name,
		contentType: 'inline',
		expectedValue: processTextNode(props.expectedValue),
		actualValue: props.actualValue,
		comparator: props.comparator,
		status: props.status,
	});
};

const isPropChildren = (children: SingleNode[]): children is PropertyNode[] =>
	children.every(child => child.type === 'prop');

const isCodeBlockNode = (item: JSX.SmstNode): item is CodeBlockNode =>
	!Array.isArray(item) && typeof item === 'object' && item !== null && item.type === 'code-block';

const isTestLineNodeArray = (children: SingleNode[]): children is TestLineNode[] =>
	children.every(child => child.type === 'test-line');

/* istanbul ignore next */
const assertUnknownIntrinsicElementType = (type: never): never => {
	throw new Error(`Unknown intrinsic element type ${type}`);
};

/* istanbul ignore next */
export const assertUnknownSectionNode = (node: never): never => {
	throw new Error('Unknown node type: ' + JSON.stringify(node));
};

/* istanbul ignore next */
export const assertUnknownTextNode = (node: never): never => {
	throw new Error('Unknown plain text node: ' + JSON.stringify(node));
};

/**
 * A factory for JSX elements. Maps React-style JSX to unist-style nodes.
 * "any" is used for children and props because TypeScript can't infer JSX element types
 * anyway and always returns whatever is provided in JSX.Element type/interface.
 * Some validation is added to ensure correct data at least runtime.
 */
export function jsx
	<K extends keyof JSX.ElementsProps>
	(type: K, props: JSX.ElementsProps[K], ...children: Array<JSX.ElementsChildren[K]>): Node;
export function jsx(
	type: keyof JSX.ElementsProps,
	props: JSX.ElementsProps[keyof JSX.ElementsProps],
	...children: JSX.SmstNode[]
): Node {
	// Do not include empty children, to allow for ternary operations in JSX
	// Flatten the children - to support nested arrays in JSX and manage fragments
	const processedChildren = normalizePlainChildren(flatten(children), type);

	switch (type) {
		case 'text':
		case 'subject':
		case 'input':
		case 'code':
			return processedChildren[0];
		case 'fragment':
			return processedChildren;
		case 'code-block':
			const codeBlockProps = props as JSX.ElementsProps['code-block'];
			const codeBlockChildren = processedChildren[0] as CodeBlockNode;
			const language = codeBlockProps?.language ?? 'javascript';

			// Type casting because IntrinsicElements definition would not allow anything other then string
			return ub('code-block', {language}, codeBlockChildren.value) as CodeBlockNode;
		case 'prop':
			// props can't be null because it's defined in intrinsicElements
			return processPropertyNode(props as JSX.ElementsProps['prop']);
		case 'props':
			if (isPropChildren(processedChildren)) {
				return ub('props', processedChildren);
			}

			throw new TypeError('Props can only accept Prop as its child');
		case 'condition':
		case 'test-line':
			const testLineProps = props as JSX.ElementsProps['test-line'];
			// Type casting for props because of IntrinsicElements definition
			const params: Omit<TestLineNode, 'type' | 'children'> = {
				title: processTextNode(testLineProps.title),
			};

			if (testLineProps.status) {
				params.status = testLineProps.status;
			}

			return ub(type, params, processedChildren) as ConditionNode | TestLineNode;
		case 'test-line-result':
			const testLineResultProps = props as JSX.ElementsProps['test-line-result'];
			if (isTestLineNodeArray(processedChildren) && processedChildren.length === 1) {
				return ub('test-line-result', {
					status: testLineResultProps.status,
					message: processTextNode(testLineResultProps.message),
					screenshot: props?.screenshot,
				}, processedChildren);
			}

			throw new TypeError('TestLineResult expects a single child of type TestLine');
		default:
			/* istanbul ignore next */
			return assertUnknownIntrinsicElementType(type);
	}
}
