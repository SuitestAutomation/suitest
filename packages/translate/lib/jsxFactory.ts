/// <reference path="../types/intrinsicElements.d.ts" />
/// <reference path="../types/unistTestLine.d.ts" />
import ub from 'unist-builder';

const plainTypes = ['text', 'code', 'bold', 'emphasis'];
const isTextNode = (input: SingleNode): input is TextNode => plainTypes.includes(input.type);

type DeepArrayOrOne<T> = T | Array<T | DeepArrayOrOne<T>>;
export const flatten = <T extends any>(input: DeepArrayOrOne<T>): T[] => {
	if (Array.isArray(input)) {
		return input.reduce((acc: T[], item: DeepArrayOrOne<T>) => acc.concat(flatten(item)), []);
	}

	return [input];
};

/**
 * Normalize textual children - make them flat and merge sequential text items of same type
 */
const normalizePlainChildren = (
	children: Array<SingleNode | string>,
	parentType: keyof JSX.IntrinsicElements
): SingleNode[] =>
	children.reduce((output: SingleNode[], child: SingleNode | string) => {
		if (typeof child === 'string') {
			// Wrap plain strings into text nodes
			child = ub(plainTypes.includes(parentType) ? parentType : 'text', child) as PlainTextNode;
		}

		// Nested child is a textual node
		const lastChild = output[output.length - 1];

		if (isTextNode(child) && lastChild?.type === child.type) {
			// Textual node with same time as the previous one - merge them
			output.splice(output.length - 1, 1, {
				type: child.type,
				value: lastChild.value + child.value,
			});
		} else {
			output.push(child);
		}

		return output;
	}, []);

const processLabel = (label?: string | SingleNode | Array<string | SingleNode>): SingleNode[] | undefined => {
	if (typeof label !== 'undefined') {
		return normalizePlainChildren(flatten([label]), 'text');
	}

	return undefined;
};

const isCellChildren = (children: SingleNode[]): children is CellNode[] =>
	children.every(child => child.type === 'cell');

const isRowChildren = (children: SingleNode[]): children is RowNode[] =>
	children.every(child => child.type === 'row');

const isDictionaryRowChildren = (children: SingleNode[]): children is DictionaryRowNode[] =>
	children.every(child => child.type === 'row' && child.children.length === 2);

const isTextChildren = (children: SingleNode[]): children is TextNode[] =>
	children.every(isTextNode);

/* istanbul ignore next */
const assertUnknownIntrinsicElementType = (type: never): never => {
	throw new Error(`Unknown intrinsic element type ${type}`);
};

/* istanbul ignore next */
export const assertUnknownSectionNode = (node: never): never => {
	throw new Error('Unknown node type: ' + JSON.stringify(node));
};

/* istanbul ignore next */
export const assertUnknownTextNode = (node: never): never => {
	throw new Error('Unknown plain text node: ' + JSON.stringify(node));
};

/**
 * A factory for JSX elements. Maps React-style JSX to unist-style nodes.
 * "any" is used for children and props because TypeScript can't infer JSX element types
 * anyway and always returns whatever is provided in JSX.Element type/interface.
 * Some validation is added to ensure correct data at least runtime.
 */
export const jsx = (
	type: keyof JSX.IntrinsicElements,
	props: {[key: string]: any} | null,
	...children: Array<Node | string>
): Node => {
	// Do not include empty children, to allow for ternary operations in JSX
	// Flatten the children - to support nested arrays in JSX and manage fragments
	const processedChildren = normalizePlainChildren(flatten(children.filter(Boolean)), type);

	switch (type) {
		case 'text': // TODO nyc for some reason reports an uncovered branch here
		case 'bold':
		case 'emphasis':
		case 'code':
			return processedChildren[0];
		case 'fragment':
			return processedChildren;
		case 'cell':
			if (isTextChildren(processedChildren)) {
				return ub('cell', processedChildren);
			}

			throw new TypeError('Cell can only accept section or textual nodes');
		case 'row':
			if (isCellChildren(processedChildren)) {
				return ub('row', processedChildren);
			}

			throw new TypeError('Row can only accept Cell as its child');
		case 'table':
			if (isRowChildren(processedChildren)) {
				return ub('table', {label: processLabel(props?.label)}, processedChildren);
			}

			throw new TypeError('Table can only accept Row as its child');
		case 'dictionary':
			if (isDictionaryRowChildren(processedChildren)) {
				return ub('dictionary', {label: processLabel(props?.label)}, processedChildren);
			}

			throw new TypeError('Dictionary can only accept Row with 2 Cells as its child');
		case 'paragraph':
			return ub(type, processedChildren) as ParagraphNode;
		case 'condition':
		case 'test-line':
			// Type casting for props because of IntrinsicElements definition
			return ub(type, {
				title: processLabel((props as {title: any}).title),
			}, processedChildren) as ConditionNode | TestLineNode;
		case 'code-block':
			// Type casting because IntrinsicElements definition would not allow anything other then string
			return ub('code-block', {label: processLabel(props?.label)}, (processedChildren[0] as TextNode).value);
		case 'alert':
			// Type casting for level is done because we can be sure it's always provided - it's enforced in
			// IntrinsicElements definition
			return ub('alert', {level: (props as {level: any}).level}, processedChildren) as AlertNode;
		default:
			/* istanbul ignore next */
			return assertUnknownIntrinsicElementType(type);
	}
};
