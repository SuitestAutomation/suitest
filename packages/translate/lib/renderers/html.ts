/// <reference path="../../types/unistTestLine.d.ts" />
import {assertUnknownSectionNode} from '../jsxFactory';
import {escapeHtml} from '../utils';

/**
 * Helper function to build HTML figure element
 * @param className - class name to be added to figure
 * @param content - body content of the figure, already renderer to string
 * @param titleNodes - optional figcaption
 * @param prefix - text to insert before figcaption
 */
const renderFigure = (className: string, content: string, titleNodes?: TextNode[], prefix = ''): string => {
	const output = [`<figure class="${className}">`];

	if (titleNodes?.length) {
		output.push(`<figcaption>${prefix + titleNodes.map(renderNode).join('')}</figcaption>`);
	}

	output.push(content, '</figure>');

	return output.join('');
};

const renderHtmlCodeBlockNode = (node: CodeBlockNode): string => renderFigure(
	'suitest-test-line__code-block',
	`<pre><code>${node.value}</code></pre>`,
	node.label
);

const renderHtmlDictionaryNode = (node: DictionaryNode): string => renderFigure(
	'suitest-test-line__dictionary',
	[
		'<dl>',
		...node.children.map(row => ([
			'<dt>',
			...row.children[0].children.map(renderNode),
			'</dt><dd>',
			...row.children[1].children.map(renderNode),
			'</dd>',
		]).join('')),
		'</dl>',
	].join(''),
	node.label
);

const renderHtmlTableNode = (node: TableNode): string => renderFigure(
	'suitest-test-line__table',
	[
		'<table>',
		...node.children.map(row => ([
			'<tr>',
			...row.children.map(cell => '<td>' + cell.children.map(renderNode).join('') + '</td>'),
			'</tr>',
		].join(''))),
		'</table>',
	].join(''),
	node.label
);

const renderHtmlConditionNode = (node: ConditionNode): string => renderFigure(
	'suitest-test-line__condition',
	node.children.map(renderNode).join(''),
	node.title,
	'Condition: '
);

const renderHtmlTestLineNode = (node: TestLineNode): string => renderFigure(
	'suitest-test-line',
	node.children.map(renderNode).join(''),
	node.title
);

const renderNode = (node: SingleNode): string => {
	switch (node.type) {
		case 'text':
			return escapeHtml(node.value ?? '');
		case 'code':
			return `<code class="suitest-test-line__text--code">${node.value ?? ''}</code>`;
		case 'bold':
		case 'emphasis':
			return `<span class="suitest-test-line__text--${node.type}">${escapeHtml(node.value ?? '')}</span>`;
		case 'paragraph':
			return `<p class="suitest-test-line__paragraph">${node.children.map(renderNode).join('')}</p>`;
		case 'code-block':
			return renderHtmlCodeBlockNode(node);
		case 'dictionary':
			return renderHtmlDictionaryNode(node);
		case 'table':
			return renderHtmlTableNode(node);
		case 'test-line':
			return renderHtmlTestLineNode(node);
		case 'condition':
			return renderHtmlConditionNode(node);
		case 'row':
		case 'cell':
			throw new Error('Cell and Row nodes can only be rendered as part of table or dictionary');
		case 'alert':
			// TODO implement
			return `<p class="suitest-test-line__paragraph">${node.children.map(renderNode).join('')}</p>`;
		default:
			/* istanbul ignore next */
			return assertUnknownSectionNode(node);
	}
};

export const toHtml = (node: Node): string => {
	if (!Array.isArray(node)) {
		node = [node];
	}

	return node.map(renderNode).join('');
};
