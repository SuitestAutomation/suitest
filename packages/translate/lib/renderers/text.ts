/// <reference path="../../types/unistTestLine.d.ts" />
import {assertUnknownSectionNode, assertUnknownTextNode} from '../jsxFactory';

type RenderTextFunc = (node: TextNode) => string;

const nl = '\n';
const tab = '  ';

const format = {
	cancel: '\u001b[0m',
	green: '\u001b[32m',
	cyan: '\u001b[36m',
	underscore: '\u001b[4m',
};

/**
 * Render a single text node as plain text
 */
const renderPlainTextNode: RenderTextFunc = (node: TextNode): string => node.value;

/**
 * Render a single text node with TTI colors
 */
const renderFormattedTextNode: RenderTextFunc = (node: TextNode): string => {
	const value = node.value ?? '';

	switch (node.type) {
		case 'text':
			return value;
		case 'bold':
			return format.green + value + format.cancel;
		case 'emphasis':
			return format.underscore + value + format.cancel;
		case 'code':
			return format.cyan + value + format.cancel;
		default:
			return assertUnknownTextNode(node);
	}
};

/**
 * Split a single AST node into several based on provided length of the first node
 */
const splitNode = (textNode: TextNode, length: number): [TextNode, TextNode] => ([
	{
		...textNode,
		value: textNode.value.slice(0, length),
	},
	{
		...textNode,
		value: textNode.value.slice(length),
	},
]);

/**
 * Wrapped textual nodes to fit max length
 * @TODO consider wrapping by word and not by any character
 */
const wrappedTextNodes = (textNodes: TextNode[], maxLineLength = 60): [number, TextNode[][]] => {
	// Keep function immutable
	textNodes = [...textNodes];

	const output: TextNode[][] = [[]];
	let maxActualLength = 0;
	let currentLineLength = 0;

	while (textNodes.length) {
		const firstNode = textNodes.shift()!;

		if (firstNode.value.length < maxLineLength - currentLineLength) {
			// The whole textNode can fit into the line
			currentLineLength += firstNode.value.length;
			output[output.length - 1].push(firstNode);
			maxActualLength = currentLineLength > maxActualLength ? currentLineLength : maxActualLength;
		} else {
			// Need to split text node
			const splitNodes = splitNode(firstNode, maxLineLength - currentLineLength);
			currentLineLength = 0;
			output[output.length - 1].push(splitNodes[0]);
			output.push([]);
			textNodes.splice(0, 1, splitNodes[1]);
			maxActualLength = maxLineLength;
		}
	}

	return [maxActualLength, output];
};

const renderDictionary = (node: DictionaryNode, renderTextNode: RenderTextFunc, prefix = ''): string => {
	const label = node.label?.length ? nl + prefix + node.label.map(renderTextNode).join('') + nl : '';

	const renderedRowsWithMaxLength = node.children
		.map(row => row.children.map(cell => wrappedTextNodes(cell.children)));

	const maxTermLength = renderedRowsWithMaxLength
		.reduce<number>((acc, [[termLength]]) => acc > termLength ? acc : termLength, 0);

	return label + renderedRowsWithMaxLength.map(([[, term], [, definition]]) => {
		const lines: string[] = [];

		while (term.length || definition.length) {
			lines.push([
				prefix,
				(term.shift() ?? []).map(renderTextNode).join('').padStart(maxTermLength),
				(lines.length ? '  ' : ': '),
				(definition.shift() ?? []).map(renderTextNode).join(''),
			].join(''));
		}

		return lines.join(nl);
	}).join(nl) + nl;
};

const renderTable = (node: TableNode, renderTextNode: RenderTextFunc, prefix = ''): string => {
	const label = node.label?.length ? nl + prefix + node.label.map(renderTextNode).join('') + nl : '';

	const renderedRowsWithMaxLength = node.children
		.map(row => row.children.map(cell => wrappedTextNodes(cell.children)));

	const columnWidth = renderedRowsWithMaxLength.reduce<number[]>((acc, row) => {
		row.forEach(([cellLength], i) => {
			if (cellLength > (acc[i] ?? 0)) {
				acc[i] = cellLength;
			}
		});

		return acc;
	}, []);

	const rows: string[] = [];

	renderedRowsWithMaxLength.map(row => {
		const rowLines = [];

		while (row.some(cell => cell[1].length)) {
			rowLines.push([
				prefix,
				'│ ',
				row.map(([, cell], i) => (cell.shift() ?? []).map(renderTextNode).join('').padEnd(columnWidth[i])).join(' │ '),
				' │',
			].join(''));
		}

		rows.push(rowLines.join(nl));
	});

	return label + prefix + '┌' + columnWidth.map(column => ''.padStart(column + 2, '─')).join('┬') + '┐' + nl +
		rows.join('\n' + prefix + '├' + columnWidth.map(column => ''.padStart(column + 2, '─')).join('┼') + '┤' + '\n') +
		nl + prefix + '└' + columnWidth.map(column => ''.padStart(column + 2, '─')).join('┴') + '┘';
};

const renderCodeBlock = (node: CodeBlockNode, renderTextNode: RenderTextFunc, prefix = ''): string => {
	let codeBlockPrefix = prefix;
	let label = '';

	if (node.label?.length) {
		label = nl + prefix + node.label.map(renderTextNode).join('');
		codeBlockPrefix += tab;
	}

	return label + nl + node.value.split(nl).map(line => codeBlockPrefix + '> ' + line).join(nl) + nl;
};

const renderNode = (node: SingleNode, renderTextNode: RenderTextFunc, prefix = ''): string => {
	switch (node.type) {
		case 'text':
		case 'code':
		case 'emphasis':
		case 'bold':
			return renderTextNode(node);
		case 'paragraph':
			return prefix + node.children.map(renderTextNode).join('') + nl;
		case 'code-block':
			return renderCodeBlock(node, renderTextNode, prefix);
		case 'dictionary':
			return renderDictionary(node, renderTextNode, prefix);
		case 'table':
			return renderTable(node, renderTextNode, prefix);
		case 'test-line':
			return renderTestLineOrCondition(node, renderTextNode);
		case 'condition':
			return nl + prefix + 'Condition: ' + renderTestLineOrCondition(node, renderTextNode, prefix + tab);
		case 'cell':
		case 'row':
			throw new Error('Cell and Row nodes can only be rendered as part of table or dictionary');
		default:
			return assertUnknownSectionNode(node);
	}
};

const renderTestLineOrCondition = (
	node: TestLineNode | ConditionNode,
	renderTextNode: RenderTextFunc,
	prefix = ''
): string => {
	const title = node.title.map(renderTextNode).join('');
	const body = node.children.map(child => renderNode(child, renderTextNode, prefix + tab)).join('');

	return title + nl + body;
};

export const toText = (node: Node, format = true): string => {
	const renderTextNode = format ? renderFormattedTextNode : renderPlainTextNode;

	if (!Array.isArray(node)) {
		node = [node];
	}

	return node.map(n => renderNode(n, renderTextNode)).join('');
};
