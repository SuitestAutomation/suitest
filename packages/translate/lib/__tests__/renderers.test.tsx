import {jsx} from '../jsxFactory';
import {toText} from '../renderers/text';
import {toHtml} from '../renderers/html';

describe('AST renderers', () => {
	const plainText = <fragment>TEXT</fragment>;
	const boldText = <bold>BOLD</bold>;
	const emphText = <emphasis>EMPH</emphasis>;
	const codeText = <code>CODE</code>;
	const paragraph = <paragraph>TEST <bold>bold</bold></paragraph>;

	const simpleCodeBlockWithLabel = <code-block label="The code">
		someJS();
	</code-block>;
	const longCodeBlock = <code-block>
		someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS();
	</code-block>;

	const simpleTableWithLabel = <table label="Simple label">
		<row>
			<cell>1:1</cell>
			<cell>1:2</cell>
		</row>
		<row>
			<cell>2:1</cell>
			<cell>2:2</cell>
		</row>
	</table>;
	const longTable = <table>
		<row>
			<cell>1:1</cell>
			<cell>1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2</cell>
		</row>
	</table>;

	const simpleDictionaryWithLabel = <dictionary label={<bold>BOLD label</bold>}>
		<row>
			<cell>Term</cell>
			<cell>Definition</cell>
		</row>
		<row>
			<cell>Longer Term 2</cell>
			<cell>Definition 2</cell>
		</row>
	</dictionary>;
	const longDictionary = <dictionary>
		<row>
			<cell>Term</cell>
			<cell>Definition Definition Definition Definition Definition Definition Definition Definition</cell>
		</row>
	</dictionary>;

	const condition = <condition title="I am condition">
		<paragraph>I am content</paragraph>
	</condition>;
	const testLine = <test-line title="I am test line">
		<paragraph>I am content</paragraph>
	</test-line>;

	describe('plain text renderer', () => {
		it('should handle textual nodes', () => {
			expect(toText(plainText, false)).toEqual('TEXT');
			expect(toText(boldText, false)).toEqual('BOLD');
			expect(toText(emphText, false)).toEqual('EMPH');
			expect(toText(codeText, false)).toEqual('CODE');
			expect(toText(paragraph, false)).toEqual('TEST bold\n');
		});

		it('should handle code blocks', () => {
			expect(toText(simpleCodeBlockWithLabel, false)).toMatchSnapshot();
			expect(toText(longCodeBlock, false)).toMatchSnapshot();
		});

		it('should handle tables', () => {
			expect(toText(simpleTableWithLabel, false)).toMatchSnapshot();
			expect(toText(longTable, false)).toMatchSnapshot();
		});

		it('should handle dictionaries', () => {
			expect(toText(simpleDictionaryWithLabel, false)).toMatchSnapshot();
			expect(toText(longDictionary, false)).toMatchSnapshot();
		});

		it('should render condition', () => {
			expect(toText(condition, false)).toMatchSnapshot();
		});

		it('should render test line', () => {
			expect(toText(testLine, false)).toMatchSnapshot();
		});
	});

	describe('formatted text renderer', () => {
		it('should handle textual nodes', () => {
			expect(toText(plainText, true)).toEqual('TEXT');
			expect(toText(boldText, true)).toEqual('\u001b[32mBOLD\u001b[0m');
			expect(toText(emphText, true)).toEqual('\u001b[4mEMPH\u001b[0m');
			expect(toText(codeText, true)).toEqual('\u001b[36mCODE\u001b[0m');
			expect(toText(paragraph, true)).toEqual('TEST \u001b[32mbold\u001b[0m\n');
		});

		it('should handle code blocks', () => {
			expect(toText(simpleCodeBlockWithLabel, true)).toMatchSnapshot();
			expect(toText(longCodeBlock, true)).toMatchSnapshot();
		});

		it('should handle tables', () => {
			expect(toText(simpleTableWithLabel, true)).toMatchSnapshot();
			expect(toText(longTable, true)).toMatchSnapshot();
		});

		it('should handle dictionaries', () => {
			expect(toText(simpleDictionaryWithLabel, true)).toMatchSnapshot();
			expect(toText(longDictionary, true)).toMatchSnapshot();
		});

		it('should render condition', () => {
			expect(toText(condition, true)).toMatchSnapshot();
		});

		it('should render test line', () => {
			expect(toText(testLine, true)).toMatchSnapshot();
		});
	});

	describe('html renderer', () => {
		it('should handle textual nodes', () => {
			expect(toHtml(plainText)).toEqual('TEXT');
			expect(toHtml(boldText)).toEqual('<span class="suitest-test-line__text--bold">BOLD</span>');
			expect(toHtml(emphText)).toEqual('<span class="suitest-test-line__text--emphasis">EMPH</span>');
			expect(toHtml(codeText)).toEqual('<code class="suitest-test-line__text--code">CODE</code>');
			expect(toHtml(paragraph)).toMatchSnapshot();
		});

		it('should handle code blocks', () => {
			expect(toHtml(simpleCodeBlockWithLabel)).toMatchSnapshot();
			expect(toHtml(longCodeBlock)).toMatchSnapshot();
		});

		it('should handle tables', () => {
			expect(toHtml(simpleTableWithLabel)).toMatchSnapshot();
			expect(toHtml(longTable)).toMatchSnapshot();
		});

		it('should handle dictionaries', () => {
			expect(toHtml(simpleDictionaryWithLabel)).toMatchSnapshot();
			expect(toHtml(longDictionary)).toMatchSnapshot();
		});

		it('should render condition', () => {
			expect(toHtml(condition)).toMatchSnapshot();
		});

		it('should render test line', () => {
			expect(toHtml(testLine)).toMatchSnapshot();
		});
	});
});
