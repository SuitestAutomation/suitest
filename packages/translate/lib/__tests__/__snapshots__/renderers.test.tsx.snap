// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AST renderers formatted text renderer should handle code blocks 1`] = `
"
The code
  > someJS();
"
`;

exports[`AST renderers formatted text renderer should handle code blocks 2`] = `
"
> someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS();
"
`;

exports[`AST renderers formatted text renderer should handle dictionaries 1`] = `
"
[32mBOLD label[0m
         Term: Definition
Longer Term 2: Definition 2
"
`;

exports[`AST renderers formatted text renderer should handle dictionaries 2`] = `
"Term: Definition Definition Definition Definition Definition Defin
      ition Definition Definition
"
`;

exports[`AST renderers formatted text renderer should handle tables 1`] = `
"
Simple label
┌─────┬─────┐
│ 1:1 │ 1:2 │
├─────┼─────┤
│ 2:1 │ 2:2 │
└─────┴─────┘
"
`;

exports[`AST renderers formatted text renderer should handle tables 2`] = `
"┌─────┬──────────────────────────────────────────────────────────────┐
│ 1:1 │ 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2  │
│     │ 1:2 1:2 1:2                                                  │
└─────┴──────────────────────────────────────────────────────────────┘
"
`;

exports[`AST renderers formatted text renderer should render condition 1`] = `
"
Condition: I am condition
    I am content
"
`;

exports[`AST renderers formatted text renderer should render test line 1`] = `
"I am test line
  I am content
"
`;

exports[`AST renderers html renderer should handle code blocks 1`] = `"<figure class=\\"suitest-test-line__code-block\\"><figcaption>The code</figcaption><pre><code>someJS();</code></pre></figure>"`;

exports[`AST renderers html renderer should handle code blocks 2`] = `"<figure class=\\"suitest-test-line__code-block\\"><pre><code>someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS();</code></pre></figure>"`;

exports[`AST renderers html renderer should handle dictionaries 1`] = `"<figure class=\\"suitest-test-line__dictionary\\"><figcaption><span class=\\"suitest-test-line__text--bold\\">BOLD label</span></figcaption><dl><dt>Term</dt><dd>Definition</dd><dt>Longer Term 2</dt><dd>Definition 2</dd></dl></figure>"`;

exports[`AST renderers html renderer should handle dictionaries 2`] = `"<figure class=\\"suitest-test-line__dictionary\\"><dl><dt>Term</dt><dd>Definition Definition Definition Definition Definition Definition Definition Definition</dd></dl></figure>"`;

exports[`AST renderers html renderer should handle tables 1`] = `"<figure class=\\"suitest-test-line__table\\"><figcaption>Simple label</figcaption><table><tr><td>1:1</td><td>1:2</td></tr><tr><td>2:1</td><td>2:2</td></tr></table></figure>"`;

exports[`AST renderers html renderer should handle tables 2`] = `"<figure class=\\"suitest-test-line__table\\"><table><tr><td>1:1</td><td>1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2</td></tr></table></figure>"`;

exports[`AST renderers html renderer should handle textual nodes 1`] = `"<p class=\\"suitest-test-line__paragraph\\">TEST <span class=\\"suitest-test-line__text--bold\\">bold</span></p>"`;

exports[`AST renderers html renderer should render condition 1`] = `"<figure class=\\"suitest-test-line__condition\\"><figcaption>Condition: I am condition</figcaption><p class=\\"suitest-test-line__paragraph\\">I am content</p></figure>"`;

exports[`AST renderers html renderer should render test line 1`] = `"<figure class=\\"suitest-test-line\\"><figcaption>I am test line</figcaption><p class=\\"suitest-test-line__paragraph\\">I am content</p></figure>"`;

exports[`AST renderers plain text renderer should handle code blocks 1`] = `
"
The code
  > someJS();
"
`;

exports[`AST renderers plain text renderer should handle code blocks 2`] = `
"
> someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS(); someJS();
"
`;

exports[`AST renderers plain text renderer should handle dictionaries 1`] = `
"
BOLD label
         Term: Definition
Longer Term 2: Definition 2
"
`;

exports[`AST renderers plain text renderer should handle dictionaries 2`] = `
"Term: Definition Definition Definition Definition Definition Defin
      ition Definition Definition
"
`;

exports[`AST renderers plain text renderer should handle tables 1`] = `
"
Simple label
┌─────┬─────┐
│ 1:1 │ 1:2 │
├─────┼─────┤
│ 2:1 │ 2:2 │
└─────┴─────┘
"
`;

exports[`AST renderers plain text renderer should handle tables 2`] = `
"┌─────┬──────────────────────────────────────────────────────────────┐
│ 1:1 │ 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2 1:2  │
│     │ 1:2 1:2 1:2                                                  │
└─────┴──────────────────────────────────────────────────────────────┘
"
`;

exports[`AST renderers plain text renderer should render condition 1`] = `
"
Condition: I am condition
    I am content
"
`;

exports[`AST renderers plain text renderer should render test line 1`] = `
"I am test line
  I am content
"
`;
